org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/parser/Parser.java

```diff
@@ -1181,37 +1174,32 @@ public RecoveredElement buildInitialRecoveryState(){
 			}
 			continue;
 		}
-		if (node instanceof ImportReference){
-			ImportReference importRef = (ImportReference) node;
+		if (node instanceof ImportReference importRef){
 			element = element.add(importRef, 0);
 			this.lastCheckPoint = importRef.declarationSourceEnd + 1;
 		}
 		if(this.statementRecoveryActivated) {
-			if(node instanceof Block) {
-				Block block = (Block) node;
+			if(node instanceof Block block) {
 				element = element.add(block, 0);
 				this.lastCheckPoint = block.sourceEnd + 1;
-			} else if(node instanceof LocalDeclaration) {
-				LocalDeclaration statement = (LocalDeclaration) node;
+			} else if(node instanceof LocalDeclaration statement) {
 				element = element.add(statement, 0);
 				this.lastCheckPoint = statement.sourceEnd + 1;
-			} else if(node instanceof Expression &&  ((Expression) node).isTrulyExpression()) {
+			} else if(node instanceof Expression statement &&  statement.isTrulyExpression()) {
```

Comment:
I think it should not be called statement if it’s an expression

Reply:
Though that was already the case prior to this cleanup :(

Reply:
I wouldn't mind changing it by the way but then it would be "changing the code otherwise" as @iloveeclipse mentioned should be avoided.
I would recommend that you open a PR that change the name and if it is merged earlier, I'll adapt this patch.